src/FirmwarePlugin/APM/APMParameterMetaData.cc

```diff
@@ -78,29 +79,396 @@ QVariant APMParameterMetaData::_stringToTypedVariant(const QString& string, Fact
+                qCDebug(APMParameterMetaDataLog) << "read Units: " << units;
+                rawMetaData->units = units;
+            }
+        } else if (elementName == "values") {
+            // doing nothing value will follow anyway. May be used for sanity checking.
+        } else if (elementName == "value") {
+            QString valueValue = xml.attributes().value("code").toString();
+            QString valueName = xml.readElementText();
+            qCDebug(APMParameterMetaDataLog) << "read value parameter " << "value desc: "
+                                             << valueName << "code: " << valueValue;
+            values << QPair<QString,QString>(valueValue, valueName);
+            rawMetaData->values = values;
+        } else {
+            qCDebug(APMParameterMetaDataLog) << "Unknown parameter element in XML: " << elementName;
+        }
+        xml.readNext();
+        elementName = xml.name().toString();
+    }
+    return true;
 }
 
 /// Override from FactLoad which connects the meta data to the fact
-void APMParameterMetaData::addMetaDataToFact(Fact* fact)
+void APMParameterMetaData::addMetaDataToFact(Fact* fact, MAV_TYPE vehicleType)
 {
     _loadParameterFactMetaData();
 
-    // FIXME: Will need to switch here based on _vehicle->firmwareType() to pull right set of meta data
+    const QString mavTypeString = mavTypeToString(vehicleType);
+    APMFactMetaDataRaw* rawMetaData = NULL;
+
+    // check if we have metadata for fact, use generic otherwise
+    if (_vehicleTypeToParametersMap[mavTypeString].contains(fact->name())) {
+        rawMetaData = _vehicleTypeToParametersMap[mavTypeString][fact->name()];
+    } else if (_vehicleTypeToParametersMap["libraries"].contains(fact->name())) {
+        rawMetaData = _vehicleTypeToParametersMap["libraries"][fact->name()];
+    }
+
+    FactMetaData *metaData = new FactMetaData(fact->type(), fact);
+
+    // we don't have data for this fact
+    if (!rawMetaData) {
+        fact->setMetaData(metaData);
+        qCWarning(APMParameterMetaDataLog) << "No metaData for " << fact->name() << "using generic metadata";
+        return;
+    }
+
+    metaData->setName(rawMetaData->name);
+    metaData->setGroup(rawMetaData->group);
+
+    if (!rawMetaData->shortDescription.isEmpty()) {
+        metaData->setShortDescription(rawMetaData->shortDescription);
+    }
+    
+    if (!rawMetaData->longDescription.isEmpty()) {
+        metaData->setLongDescription(rawMetaData->longDescription);
+    }
+
+    if (!rawMetaData->units.isEmpty()) {
+        metaData->setUnits(rawMetaData->units);
+    }
+
+
+    if (!rawMetaData->min.isEmpty()) {
+        QVariant varMin;
+        QString errorString;
+        if (metaData->convertAndValidate(rawMetaData->min, true /* convertOnly */, varMin, errorString)) {
+            metaData->setMin(varMin);
+        } else {
+            qCWarning(APMParameterMetaDataLog) << "Invalid min value, name:" << metaData->name()
```

Comment:
GIven the state of the meta data if this fails I would just qCDebug and toss min/max values.